### PR TITLE
[file_selector] Switch to Pigeon for macOS

### PR DESCRIPTION
**Pull request by stuartmorgan**
Converts from direct method channel use to the new experimental Swift generator.

Some specific changes to call out that aren't just direct conversion:
- Adjusts the API boundary slightly to map more directly to `NSOpenPanel` and `NSSavePanel`, in support of the general goal of preferring to have logic in Dart.
  - This also eliminates the confusing (and with manual method channels, error prone) special handling of the open completion handler where it returned different types (either string or list of strings) across the process boundary depending on the call mode.
- Dart unit tests are now less fragile to future changes since they can assert only the arguments the test are about, instead of having to be a full exact match for the call arguments.
- Added some Dart unit test coverage since there was no specific case just to test return handling (although it was part of a different test), and there was no coverage at all of cancelling.

This is a particularly strong case for a Pigeon conversion since we don't currently have a way to run integration tests for this plugin, so the Dart/Swift communication layer isn't tested, making the plugin prone to bugs where each side's unit tests pass, but the plugin doesn't actually work because of an API expectation mismatch. This offloads exactly that otherwise untested part to Pigeon, which has its own set of tests of that layer.

(The native unit tests are a little more verbose since we don't have a way to provide default values for data classes currently, but that's a small downside relative to the benefits.)

Part of https://github.com/flutter/flutter/issues/117844

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/file_selector/file_selector_macos/test/file_selector_macos_test.mocks.dart [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+// ignore_for_file: no_leading_underscores_for_library_prefixes
+import 'dart:async' as _i3;
+
+import 'package:file_selector_macos/src/messages.g.dart' as _i4;
+import 'package:mockito/mockito.dart' as _i1;
+
+import 'messages_test.g.dart' as _i2;
+
+// ignore_for_file: type=lint
```
Comment (tarrinneal) [inline packages/file_selector/file_selector_macos/test/file_selector_macos_test.mocks.dart:13]: Is this method preferred to single line ignores?

File: packages/file_selector/file_selector_macos/lib/src/messages.g.dart [hunk 2]
```diff
@@ -0,0 +1,227 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (tarrinneal) [inline packages/file_selector/file_selector_macos/lib/src/messages.g.dart:1]: Is there any reason we can't use the same gen file for all 3 locations?

**Review by tarrinneal — COMMENTED** [review]
This all seems good to me. I want to have @cbracken give final approval though.

File: packages/file_selector/file_selector_macos/test/file_selector_macos_test.mocks.dart (hunk 1 above)
Comment (stuartmorgan) [reply packages/file_selector/file_selector_macos/test/file_selector_macos_test.mocks.dart:13]: File-level ignores are pretty common in generated code. This is Mockito-generated code though so we don't control how ignores are handled (unless we hand-edit the generated code, which we avoid when possible).

File: packages/file_selector/file_selector_macos/lib/src/messages.g.dart (hunk 2 above)
Comment (stuartmorgan) [reply packages/file_selector/file_selector_macos/lib/src/messages.g.dart:1]: What do you mean "all three locations"?

File: packages/file_selector/file_selector_macos/test/file_selector_macos_test.mocks.dart (hunk 1 above)
Comment (tarrinneal) [reply packages/file_selector/file_selector_macos/test/file_selector_macos_test.mocks.dart:13]: Ah, didn't realize this was generated code.

**Review by tarrinneal — APPROVED** [review]

**Pull request merged**